### PR TITLE
 Go bindings

### PR DESCRIPTION
**Pull request by bcebere**
- [x] Add C interface for the library.
- [x] Add tests and benchmarks for the C interface.
- [x] Add Go bindings for the C interface using cgo.
- [x] Add tests and benchmarks for the Go implementation.
- [x] Add Bazel support

### CODE REVIEW

**Comment by bcebere** [comment]
Test&benchmarks output

`ok  	openmined/psi-cardinality/bloomfilter	11.499s	coverage: 100.0% of statements
ok  	openmined/psi-cardinality/client	5.869s	coverage: 93.1% of statements
ok  	openmined/psi-cardinality/server	5.265s	coverage: 93.5% of statements
PASS
ok  	openmined/psi-cardinality/bloomfilter	7.442s
goos: linux
goarch: amd64
pkg: openmined/psi-cardinality/client
BenchmarkClientCreateRequest1-4         	    4584	   5302973 ns/op	      4584 ElementsProcessed	        48.0 RequestSize	     200 B/op	      11 allocs/op
BenchmarkClientCreateRequest10-4        	    2972	   9088543 ns/op	     29720 ElementsProcessed	       471 RequestSize	    1736 B/op	      28 allocs/op
BenchmarkClientCreateRequest100-4       	     644	  31198051 ns/op	     64400 ElementsProcessed	      4701 RequestSize	   14728 B/op	     124 allocs/op
BenchmarkClientCreateRequest1000-4      	      92	 260230638 ns/op	     92000 ElementsProcessed	     47001 RequestSize	  130764 B/op	    1030 allocs/op
BenchmarkClientCreateRequest10000-4     	       9	2459388057 ns/op	     90000 ElementsProcessed	    470001 RequestSize	 2287440 B/op	   10048 allocs/op
BenchmarkClientProcessResponse1-4       	    2232	  10837531 ns/op	      2232 ElementsProcessed	     464 B/op	      25 allocs/op
BenchmarkClientProcessResponse10-4      	    1292	  18656263 ns/op	     12920 ElementsProcessed	    2960 B/op	      46 allocs/op
BenchmarkClientProcessResponse100-4     	     219	 108044514 ns/op	     21900 ElementsProcessed	   24528 B/op	     145 allocs/op
BenchmarkClientProcessResponse1000-4    	      25	 894080397 ns/op	     25000 ElementsProcessed	  221008 B/op	    1054 allocs/op
BenchmarkClientProcessResponse10000-4   	       3	7661880467 ns/op	     30000 ElementsProcessed	 3670360 B/op	   10081 allocs/op
PASS
ok  	openmined/psi-cardinality/client	285.278s
goos: linux
goarch: amd64
pkg: openmined/psi-cardinality/server
BenchmarkServerSetup1fpr3-4            	    3118	   7874429 ns/op	     31180 ElementsProcessed	    1352 B/op	      28 allocs/op
BenchmarkServerSetup10fpr3-4           	    3028	   7984480 ns/op	     30280 ElementsProcessed	    1352 B/op	      28 allocs/op
BenchmarkServerSetup100fpr3-4          	     744	  48146196 ns/op	     74400 ElementsProcessed	   10506 B/op	     124 allocs/op
BenchmarkServerSetup1000fpr3-4         	      81	 441558916 ns/op	     81000 ElementsProcessed	   87752 B/op	    1030 allocs/op
BenchmarkServerSetup10000fpr3-4        	       7	4140072618 ns/op	     70000 ElementsProcessed	 1869569 B/op	   10049 allocs/op
BenchmarkServerSetup1fpr6-4            	    2654	   7838755 ns/op	     26540 ElementsProcessed	    1384 B/op	      28 allocs/op
BenchmarkServerSetup10fpr6-4           	    3099	   7796100 ns/op	     30990 ElementsProcessed	    1384 B/op	      28 allocs/op
BenchmarkServerSetup100fpr6-4          	     734	  31384969 ns/op	     73400 ElementsProcessed	   10760 B/op	     124 allocs/op
BenchmarkServerSetup1000fpr6-4         	      92	 260518752 ns/op	     92000 ElementsProcessed	   89800 B/op	    1030 allocs/op
BenchmarkServerSetup10000fpr6-4        	       9	2485713404 ns/op	     90000 ElementsProcessed	 1893962 B/op	   10048 allocs/op
BenchmarkServerProcessRequest1-4       	    2358	  10413450 ns/op	      2358 ElementsProcessed	        48.0 ResponseSize	     304 B/op	      17 allocs/op
BenchmarkServerProcessRequest10-4      	    1741	  13806596 ns/op	     17410 ElementsProcessed	       471 ResponseSize	    2272 B/op	      34 allocs/op
BenchmarkServerProcessRequest100-4     	     493	  54652297 ns/op	     49300 ElementsProcessed	      4701 ResponseSize	   19648 B/op	     130 allocs/op
BenchmarkServerProcessRequest1000-4    	      63	 376227280 ns/op	     63000 ElementsProcessed	     47001 ResponseSize	  179968 B/op	    1036 allocs/op
BenchmarkServerProcessRequest10000-4   	       6	4089743124 ns/op	     60000 ElementsProcessed	    470001 ResponseSize	 2762389 B/op	   10054 allocs/op
PASS`

**Comment by bcebere** [comment]
**Go benchmarks - each test ran for 20 seconds.**

Test name | Calls | Time/call | elements processed | bytes alloc/call | alloc/call | 
------------ | ------------- | ------------- | ------------- | ------------- | -------------
BenchmarkClientCreateRequest1-4 | 4584| 5302973 ns/op | 4584  | 200 B/op | 11 allocs/op
BenchmarkClientCreateRequest10-4 | 2972| 9088543 ns/op | 29720  | 1736 B/op | 28 allocs/op
BenchmarkClientCreateRequest100-4 | 644| 31198051 ns/op | 64400  | 14728 B/op | 124 allocs/op
BenchmarkClientCreateRequest1000-4 | 92| 260230638 ns/op | 92000  | 130764 B/op | 1030 allocs/op
BenchmarkClientCreateRequest10000-4 | 9| 2459388057 ns/op | 90000  | 2287440 B/op | 10048 allocs/op
BenchmarkClientProcessResponse1-4 | 2232| 10837531 ns/op | 2232  | 464 B/op | 25 allocs/op
BenchmarkClientProcessResponse10-4 | 1292| 18656263 ns/op | 12920  |2960 B/op | 46 allocs/op
BenchmarkClientProcessResponse100-4 | 219| 108044514 ns/op | 21900  | 24528 B/op | 145 allocs/op
BenchmarkClientProcessResponse1000-4 | 25| 894080397 ns/op | 25000  | 221008 B/op | 1054 allocs/op
BenchmarkClientProcessResponse10000-4 | 3| 7661880467 ns/op | 30000  |3670360 B/op | 10081 allocs/op
BenchmarkServerSetup1fpr3-4 | 3118 | 7874429 ns/op | 31180  | 1352 B/op | 28 allocs/op
BenchmarkServerSetup10fpr3-4 | 3028 | 7984480 ns/op | 30280  | 1352 B/op | 28 allocs/op
BenchmarkServerSetup100fpr3-4 | 744 | 48146196 ns/op | 74400  |10506 B/op | 124 allocs/op
BenchmarkServerSetup1000fpr3-4 | 81 | 441558916 ns/op | 81000  |87752 B/op | 1030 allocs/op
BenchmarkServerSetup10000fpr3-4 | 7 |4140072618 ns/op | 70000  |1869569 B/op | 10049 allocs/op
BenchmarkServerSetup1fpr6-4 | 2654 | 7838755 ns/op | 26540  | 1384 B/op | 28 allocs/op
BenchmarkServerSetup10fpr6-4 | 3099 | 7796100 ns/op | 30990  | 1384 B/op | 28 allocs/op
BenchmarkServerSetup100fpr6-4 | 734 | 31384969 ns/op | 73400  | 10760 B/op | 124 allocs/op
BenchmarkServerSetup1000fpr6-4 | 92 | 260518752 ns/op | 92000  | 89800 B/op | 1030 allocs/op
BenchmarkServerSetup10000fpr6-4 | 9| 2485713404 ns/op | 90000  |1893962 B/op | 10048 allocs/op
BenchmarkServerProcessRequest1-4 | 2358| 10413450 ns/op | 2358  | 304 B/op | 17 allocs/op
BenchmarkServerProcessRequest10-4 | 1741| 13806596 ns/op | 17410  | 2272 B/op | 34 allocs/op
BenchmarkServerProcessRequest100-4 | 493| 54652297 ns/op | 49300  | 19648 B/op | 130 allocs/op
BenchmarkServerProcessRequest1000-4 | 63| 376227280 ns/op | 63000  | 179968 B/op | 1036 allocs/op
BenchmarkServerProcessRequest10000-4 | 6| 4089743124 ns/op | 60000  | 2762389 B/op | 10054 allocs/op

File: psi_cardinality/cpp/psi_cardinality_client_test_c.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+#include "psi_cardinality_client_c.h"
+#include "rapidjson/document.h"
+#include "rapidjson/stringbuffer.h"
+#include "rapidjson/writer.h"
+#include "util/status_matchers.h"
+
+namespace psi_cardinality {
+namespace {
+
+class PSICardinalityClientBindingTest : public ::testing::Test {
```
Comment (schoppmp) [inline psi_cardinality/cpp/psi_cardinality_client_test_c.cpp:31]: If we move the C bindings to their own directory, these tests should also get their own `cc_test` target.

File: psi_cardinality/go/bloomfilter/bloomfilter.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	bf := new(BloomFilter)
+	bf.context = C.bloom_filter_create_from_json(C.CString(encoded))
+	if bf.context == nil {
+		return nil, errors.New("failed to init bloom filter")
+	}
+	return bf, nil
+}
+
+//Destroy the context
+func (b *BloomFilter) Destroy() error {
```
Comment (schoppmp) [inline psi_cardinality/go/bloomfilter/bloomfilter.go:42]: Would it be possible to not export Destroy, but register it as a Finalizer using [runtime.SetFinalizer](https://golang.org/pkg/runtime/#SetFinalizer)? We don't need to free external resources (DB, files), only memory, so I think the usual issues with Finalizers (that they might not run) should be fine here?

File: psi_cardinality/go/server/server.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+type PSICardinalityServer struct {
+	context C.psi_cardinality_server_ctx
+}
+
+//CreateWithNewKey returns a new PSI server
+func CreateWithNewKey() (*PSICardinalityServer, error) {
+	psiServer := new(PSICardinalityServer)
+	psiServer.context = C.psi_cardinality_server_create_with_new_key()
+	if psiServer.context == nil {
+		return nil, errors.New("failed to create server context")
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:26]: If we get the error message from C as suggested above then we could forward it here.

File: psi_cardinality/cpp/BUILD [hunk 4]
```diff
@@ -18,8 +18,8 @@ package(default_visibility = ["//visibility:public"])
 
 cc_library(
     name = "bloom_filter",
-    srcs = ["bloom_filter.cpp"],
-    hdrs = ["bloom_filter.h"],
+    srcs = ["bloom_filter.cpp", "bloom_filter_c.cpp"],
+    hdrs = ["bloom_filter.h", "bloom_filter_c.h"],
```
[the remark this answers is not in the document]
Comment (bcebere) [reply psi_cardinality/cpp/BUILD:22]: Bloom filter was more like a test binding, to get it started. I can remove those bindings.

File: psi_cardinality/cpp/psi_cardinality_client_c.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    size_t len = value.size() + 1;
+    *out = new char[len];
+    strncpy(*out, value.c_str(), len);
+    *out_len = len;
+
+    return 0;
+}
+
+void psi_cardinality_client_delete_buffer(psi_cardinality_client_ctx ctx,
```
[the remark this answers is not in the document]
Comment (bcebere) [reply psi_cardinality/cpp/psi_cardinality_client_c.cpp:52]: Go has its own management for the memory and threading, separate from C.

That means that when I access a C pointer, everything Go-related goes on a stack, and it passes the control to the C env/kernel.
Go never actually gets access to the C pointers. it has the pointer concept for the garbage collector, but no pointer arithmetics, for example.

TLDR: Go doesn't actually control the memory allocated in C, so it has to be freed from C. I couldn't find a prettier solution.

[the remark this answers is not in the document]
Comment (schoppmp) [reply psi_cardinality/cpp/psi_cardinality_client_c.cpp:52]: I agree that you need to free the memory from C, but I thought this could be done easier by constructing the buffer using `malloc` instead of `new`, and then calling `C.free(out)`instead of `C.psi_cardinality_client_delete_buffer(c.context, &out)` [here](https://github.com/OpenMined/psi-cardinality/pull/12/files#diff-c6c24f591a29c4cd410bb99331d22b44R56).

[the remark this answers is not in the document]
Comment (bcebere) [reply psi_cardinality/cpp/psi_cardinality_client_c.cpp:52]: Ah, I understand it now. 

I thought it was better to make no assumptions about the memory in Go(malloc vs new), and to manage it only in one place.

File: psi_cardinality/go/bloomfilter/bloomfilter.go (hunk 2 above)
Comment (bcebere) [reply psi_cardinality/go/bloomfilter/bloomfilter.go]: I removed the bloomfilter bindings

Comment (bcebere) [reply psi_cardinality/go/bloomfilter/bloomfilter.go]: Fixed for the client/server allocations

File: psi_cardinality/c/BUILD [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "psi_cardinality_client_c",
+    srcs = ["psi_cardinality_client_c.cpp", "internal_utils.h", "internal_utils.cpp"],
```
Comment (schoppmp) [inline psi_cardinality/c/BUILD:21]: Since internal_utils is used in both the server and the client, I think we should move it to its own cc_library. You can set `visibility = ["//visibility:private"]` to ensure it's not accessible from the outside.

File: psi_cardinality/c/internal_utils.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef PSI_CARDINALITY_INTERNAL_UTILS_H_
+#define PSI_CARDINALITY_INTERNAL_UTILS_H_
+
+#include "util/statusor.h"
+
+int generate_error(private_join_and_compute::Status status, char **error_out);
```
Comment (schoppmp) [inline psi_cardinality/c/internal_utils.h:22]: Since this is only included from C++ files, let's put it in an internal namespace, e.g., `psi_cardinality::c_bindings_internal`

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  int64_t elements_processed = 0;
+  for (auto _ : state) {
+    char *server_setup = nullptr;
+    size_t server_setup_buff_len = 0;
+    psi_cardinality_server_create_setup_message(
+        server_, fpr, num_client_inputs, inputs.data(), inputs.size(),
+        &server_setup, &server_setup_buff_len);
+
+    ::benchmark::DoNotOptimize(
+        std::string(server_setup, server_setup_buff_len));
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:33]: This creates an unnecessary copy of the string. The following should do the trick without copying.
```C++
::benchmark::DoNotOptimize(server_setup);
::benchmark::ClobberMemory();
```
See also: https://github.com/google/benchmark/#preventing-optimization

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 9]
```diff
@@ -0,0 +1,191 @@
+#include "absl/strings/str_cat.h"
+#include "benchmark/benchmark.h"
+#include "psi_cardinality_client_c.h"
+#include "psi_cardinality_server_c.h"
+
+namespace psi_cardinality {
+namespace {
+
+void BM_ServerBindingSetup(benchmark::State &state, double fpr) {
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:9]: The name can be the same as in the C++ benchmarks, i.e., BM_ServerSetup. The target path will make clear whether we mean C or C++ benchmarks.

File: psi_cardinality/c/BUILD [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "psi_cardinality_client_c",
```
Comment (schoppmp) [inline psi_cardinality/c/BUILD:20]: The `_c` suffix isn't necessary since the C bindings are now in their own directory.

Comment (schoppmp) [reply psi_cardinality/c/BUILD:20]: I actually think we can also remove the `psi_cardinality_` prefix and just leave it at `client` and `server`. But let's save that part for another PR.

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+// Range is for the number of inputs, and the captured argument is the false
+// positive rate for 10k client queries.
+BENCHMARK_CAPTURE(BM_ServerBindingSetup, fpr = 0.001, 0.001)
+    ->RangeMultiplier(100)
+    ->Range(1, 1000000);
+BENCHMARK_CAPTURE(BM_ServerBindingSetup, fpr = 0.000001, 0.000001)
+    ->RangeMultiplier(100)
+    ->Range(1, 1000000);
+
+void BM_ClientBindingCreateRequest(benchmark::State &state) {
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:53]: Fix name (see above)

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+  }
+  std::string request;
+  int64_t elements_processed = 0;
+  for (auto _ : state) {
+    char *client_request = {0};
+    size_t req_len = 0;
+    psi_cardinality_client_create_request(client_, inputs.data(), inputs.size(),
+                                          &client_request, &req_len);
+
+    ::benchmark::DoNotOptimize(std::string(client_request, req_len));
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:74]: Unnecessary copy (see above)

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+  state.counters["RequestSize"] = benchmark::Counter(
+      static_cast<double>(request.size()), benchmark::Counter::kDefaults,
+      benchmark::Counter::kIs1024);
+  state.counters["ElementsProcessed"] = benchmark::Counter(
+      static_cast<double>(elements_processed), benchmark::Counter::kIsRate);
+}
+// Range is for the number of inputs.
+BENCHMARK(BM_ClientBindingCreateRequest)->RangeMultiplier(10)->Range(1, 10000);
+
+void BM_ServerBindingProcessRequest(benchmark::State &state) {
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:88]: Fix name (see above)

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+  std::string response;
+
+  int64_t elements_processed = 0;
+  for (auto _ : state) {
+    char *server_response = nullptr;
+    size_t response_len = 0;
+    psi_cardinality_server_process_request(server_, {client_request, req_len},
+                                           &server_response, &response_len);
+    response = std::string(server_response, response_len);
+    ::benchmark::DoNotOptimize(response);
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:118]: Unnecessary copy (see above)

File: psi_cardinality/c/psi_cardinality_benchmark_c.cpp [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+  state.counters["ElementsProcessed"] = benchmark::Counter(
+      static_cast<double>(elements_processed), benchmark::Counter::kIsRate);
+
+  psi_cardinality_client_delete_buffer(client_, &client_request);
+}
+
+// Range is for the number of inputs.
+BENCHMARK(BM_ServerBindingProcessRequest)->RangeMultiplier(10)->Range(1, 10000);
+
+void BM_ClientBindingProcessResponse(benchmark::State &state) {
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_benchmark_c.cpp:135]: Fix name (see above)

File: psi_cardinality/c/psi_cardinality_client_c.cpp [hunk 16]
```diff
@@ -0,0 +1,73 @@
+#include "psi_cardinality_client_c.h"
+
+#include "internal_utils.h"
+#include "psi_cardinality_client.h"
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_client_c.cpp:4]: Please use the full path here, i.e., `psi_cardinality/cpp/psi_cardinality_client.h`. IDEs sometimes have problems tracking includes with relative paths that are not in the same directory, and we might want to rename the files in the future.

File: psi_cardinality/c/psi_cardinality_client_c.cpp [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+using namespace psi_cardinality;
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_client_c.cpp:6]: [Avoid `using namespace`](https://google.github.io/styleguide/cppguide.html#Namespaces).
What you could do is import the classes you want available into an anonymous namespace, i.e.,
```C++
namespace {
  using psi_cardinality::PSICardinalityClient;
}
```

File: psi_cardinality/c/psi_cardinality_client_c.cpp [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+  size_t len = value.size() + 1;
+  *out = new char[len];
+  strncpy(*out, value.c_str(), len);
+  *out_len = len;
+
+  return 0;
+}
+
+void psi_cardinality_client_delete_buffer(psi_cardinality_client_ctx ctx,
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_client_c.cpp:55]: I'm still not 100% convinced we need extra functions to delete char buffers, but let's leave it for now.

File: psi_cardinality/c/psi_cardinality_client_c.cpp [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+int64_t psi_cardinality_client_process_response(psi_cardinality_client_ctx ctx,
+                                                const char *server_setup,
+                                                const char *server_response) {
+  auto client = static_cast<PSICardinalityClient *>(ctx);
+  if (client == nullptr) {
+    return -1;
+  }
+  auto result = client->ProcessResponse(server_setup, server_response);
+  if (!result.ok()) {
+    return -1;
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_client_c.cpp:70]: Please forward the error message using `generate_error`.

File: psi_cardinality/c/psi_cardinality_client_c.cpp [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  std::vector<std::string> in;
+  for (size_t idx = 0; idx < input_len; ++idx) {
+    in.push_back(std::string(inputs[idx].buff, inputs[idx].buff_len));
+  }
+
+  auto result = client->CreateRequest(in);
+  if (!result.ok()) {
+    return -1;
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_client_c.cpp:43]: Please forward the error message using `generate_error`.

File: psi_cardinality/c/psi_cardinality_server_c.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  std::vector<std::string> in;
+  for (size_t idx = 0; idx < input_len; ++idx) {
+    in.push_back(std::string(input[idx].buff, input[idx].buff_len));
+  }
+
+  auto result = server->CreateSetupMessage(fpr, num_client_inputs, in);
+  if (!result.ok()) {
+    return -1;
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_server_c.cpp:58]: Please forward the error message using `generate_error`.

File: psi_cardinality/c/psi_cardinality_server_c.cpp [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+                                           char **output, size_t *output_len) {
+  auto server = static_cast<PSICardinalityServer *>(ctx);
+  if (server == nullptr) {
+    return -1;
+  }
+
+  auto result = server->ProcessRequest(
+      std::string(client_request.buff, client_request.buff_len));
+  if (!result.ok()) {
+    return -1;
```
Comment (schoppmp) [inline psi_cardinality/c/psi_cardinality_server_c.cpp:81]: Please forward the error message using `generate_error`.

File: psi_cardinality/go/client/client.go [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+func Create() (*PSICardinalityClient, error) {
+	psiClient := new(PSICardinalityClient)
+
+	var err *C.char
+	ret := C.psi_cardinality_client_create(&psiClient.context, &err)
+	if ret != 0 {
+		return nil, fmt.Errorf("failed to create client context: %v(%v)", C.GoString(err), ret)
+	}
+	if psiClient.context == nil {
+		return nil, errors.New("failed to create client context: null")
```
Comment (schoppmp) [inline psi_cardinality/go/client/client.go:28]: Can this ever happen if there isn't an implementation error in our library? If not, maybe make this explicit, e.g. `errors.New("failed to create client context: Context is NULL. This should never happen")`

File: psi_cardinality/go/client/client.go [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+	if result < 0 {
+		return 0, errors.New("process response failed")
+	}
+	return int64(result), nil
+}
+
+func destroy(c *PSICardinalityClient) error {
+	if c.context == nil {
+		return errors.New("invalid context")
```
Comment (schoppmp) [inline psi_cardinality/go/client/client.go:80]: Coming to think of it, if we do nothing in case `c.context == nil` it doesn't hurt if Destroy is run multiple times, right? In that case let's maybe export it after all, because users might want to free the memory before the Finalizer is run. We can mention in the documentation that it is possible but not strictly necessary to call it.

File: psi_cardinality/go/client/client.go [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+//ProcessResponse returns the intersection size
+func (c *PSICardinalityClient) ProcessResponse(serverSetup, serverResponse string) (int64, error) {
+	if c.context == nil {
+		return 0, errors.New("invalid context")
+	}
+
+	result := C.psi_cardinality_client_process_response(c.context, C.CString(serverSetup), C.CString(serverResponse))
+
+	if result < 0 {
+		return 0, errors.New("process response failed")
```
Comment (schoppmp) [inline psi_cardinality/go/client/client.go:73]: Forward error message once it's returned from C.

File: psi_cardinality/go/client/client.go [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+			buff_len: C.ulong(len(rawInput[idx])),
+		})
+	}
+
+	var out *C.char
+	var outlen C.ulong
+
+	ret := C.psi_cardinality_client_create_request(c.context, &inputs[0], C.ulong(len(inputs)), &out, &outlen)
+	if ret < 0 {
+		return "", errors.New("create request failed")
```
Comment (schoppmp) [inline psi_cardinality/go/client/client.go:54]: Forward error message once it's returned from C.

File: psi_cardinality/go/server/server.go [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+func CreateWithNewKey() (*PSICardinalityServer, error) {
+	psiServer := new(PSICardinalityServer)
+
+	var err *C.char
+	ret := C.psi_cardinality_server_create_with_new_key(&psiServer.context, &err)
+	if ret != 0 {
+		return nil, fmt.Errorf("failed to create server context: %v(%v)", C.GoString(err), ret)
+	}
+	if psiServer.context == nil {
+		return nil, errors.New("failed to create server context: null")
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:28]: Same as in Client, please make explicit if this indicates an implementation error on our side.

File: psi_cardinality/go/server/server.go [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+		buff:     C.CString(key),
+		buff_len: C.ulong(len(key)),
+	},
+		&psiServer.context, &err)
+
+	if ret != 0 {
+		return nil, fmt.Errorf("failed to create server context: %v(%v)", C.GoString(err), ret)
+	}
+	if psiServer.context == nil {
+		return nil, errors.New("failed to create server context: null")
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:50]: Same as above

File: psi_cardinality/go/server/server.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+	var out *C.char
+	var outlen C.ulong
+
+	ret := C.psi_cardinality_server_process_request(s.context, C.struct_server_buffer_t{
+		buff:     C.CString(request),
+		buff_len: C.ulong(len(request)),
+	}, &out, &outlen)
+
+	if ret < 0 {
+		return "", errors.New("process request failed")
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:97]: Forward error message once it's returned from C.

File: psi_cardinality/go/server/server.go [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+		return "", errors.New("invalid context")
+	}
+
+	var out *C.char
+	var outlen C.ulong
+
+	ret := C.psi_cardinality_server_get_private_key_bytes(s.context, &out, &outlen)
+
+	if ret < 0 {
+		return "", errors.New("get private keys failed")
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:119]: This can never happen, since the case that `s.context == nil` is already caught above.

File: psi_cardinality/go/server/server.go [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	result := C.GoString(out)
+	C.psi_cardinality_server_delete_buffer(s.context, &out)
+
+	return result, nil
+}
+
+func destroy(s *PSICardinalityServer) error {
```
Comment (schoppmp) [inline psi_cardinality/go/server/server.go:128]: Same as in Client, maybe export Destroy and allow calling it multiple times.

File: CONTRIBUTORS.md [hunk 32]
```diff
@@ -1,3 +1,4 @@
+[Bogdan Cebere](https://github.com/bcebere)
```
[the remark this answers is not in the document]
Comment (schoppmp) [reply CONTRIBUTORS.md:1]: Also please add blank lines between the names.
